File: ruby/bob/bob.rb
class Bob
	# Why does bob use proper punctuation?
	def hey(remark)
		if silenced?(remark)
			'Fine. Be that way!'
		elsif yelled_at?(remark)
			'Whoa, chill out!'
		elsif questioned?(remark)
			'Sure.'
		else
			'Whatever.'
		end
	end

	def silenced?(remark)
		remark !~ /[^\s]/
	end

	def yelled_at?(remark)
		remark =~ /[a-zA-Z]/ && remark.upcase == remark
	end

	def questioned?(remark)
		remark.end_with?('?')
	end
end

File: haskell/bob/Bob.hs
module Bob (responseFor) where

import Data.Char

responseFor :: String -> String
responseFor interaction
    | shouting         = "Whoa, chill out!"
    | questioning      = "Sure."
    | silence          = "Fine. Be that way!"
    | otherwise        = "Whatever."
    where
      silence = all isSpace interaction
      shouting = any isAlpha interaction && (all isUpper.filter isAlpha) interaction
      questioning = (not.null) interaction &&  ((== '?').last) interaction
      letters = filter isAlpha interaction

